File: RepoTransBench/target_projects/Java/C#/happyfish100_fastdfs-client-java/public_tests/TrackerClientPublicTest.cs
using Xunit;
using Moq;
using FastDFSClient;

namespace FastDFSClient.PublicTests
{
    public class TrackerClientPublicTest
    {
        [Fact]
        public void TestConstructorsAndGetErrorCodePublic()
        {
            var groupMock = new Mock<TrackerGroup>();
            var client = new TrackerClient(groupMock.Object);

            Assert.Same(groupMock.Object, client.TrackerGroup);

            var client2 = new TrackerClient();
            Assert.NotNull(client2.TrackerGroup);

            client.errno = 10;
            Assert.Equal(10, client.GetErrorCode());
        }

        [Fact]
        public void TestGetTrackerServerPublic()
        {
            var groupMock = new Mock<TrackerGroup>();
            var serverMock = new Mock<TrackerServer>().Object;
            groupMock.Setup(g => g.GetTrackerServer()).Returns(serverMock);

            var client = new TrackerClient(groupMock.Object);
            var result = client.GetTrackerServer();
            Assert.Same(serverMock, result);
        }

        [Fact]
        public void TestGetConnectionFailoverPublic()
        {
            var groupMock = new Mock<TrackerGroup>();
            var servers = new[] {
                new Mock<TrackerServer>(),
                new Mock<TrackerServer>()
            };
            var connection1 = new Mock<IConnection>().Object;
            groupMock.Setup(g => g.TrackerServers).Returns(new TrackerServer[] { servers[0].Object, servers[1].Object });
            groupMock.Setup(g => g.GetTrackerServer()).Returns(servers[0].Object);
            servers[0].Setup(s => s.GetConnection()).Throws(new System.IO.IOException("fail1"));
            servers[0].Setup(s => s.Index).Returns(0);
            groupMock.Setup(g => g.GetTrackerServer(1)).Returns(servers[1].Object);
            servers[1].Setup(s => s.GetConnection()).Returns(connection1);

            var client = new TrackerClient(groupMock.Object);
            var result = client.GetConnection(null);
            Assert.Same(connection1, result);
        }
    }
}

File: RepoTransBench/target_projects/Java/Go/happyfish100_fastdfs-client-java/public_tests/tracker_client_public_test.go
package public_tests

import (
	"io"
	"testing"

	"github.com/stretchr/testify/assert"
)

// Mocks for TrackerGroup, TrackerClient, TrackerServer, Connection interfaces (public versions)

type TrackerGroupPublic interface {
	GetTrackerServer() TrackerServerPublic
	GetTrackerServerWithIdx(int) TrackerServerPublic
	GetTrackerServers() []TrackerServerPublic
}

type TrackerServerPublic interface {
	GetConnection() ConnectionPublic
	GetIndex() int
}

type ConnectionPublic interface {
	io.Closer
}

type TrackerClientPublic struct {
	tracker_group TrackerGroupPublic
	errno         int
}

func NewTrackerClientPublic(g TrackerGroupPublic) *TrackerClientPublic {
	return &TrackerClientPublic{tracker_group: g, errno: 0}
}
func (tc *TrackerClientPublic) GetErrorCode() int {
	return tc.errno
}
func (tc *TrackerClientPublic) GetTrackerServer() TrackerServerPublic {
	return tc.tracker_group.GetTrackerServer()
}
func (tc *TrackerClientPublic) GetConnection(ts TrackerServerPublic) ConnectionPublic {
	if ts != nil {
		return ts.GetConnection()
	}
	servers := tc.tracker_group.GetTrackerServers()
	for _, srv := range servers {
		conn := srv.GetConnection()
		if conn != nil {
			return conn
		}
	}
	return nil
}

type MockTrackerGroupPublic struct {
	servers []TrackerServerPublic
	server  TrackerServerPublic
}

func (mtg *MockTrackerGroupPublic) GetTrackerServer() TrackerServerPublic {
	return mtg.server
}
func (mtg *MockTrackerGroupPublic) GetTrackerServerWithIdx(idx int) TrackerServerPublic {
	return mtg.servers[idx]
}
func (mtg *MockTrackerGroupPublic) GetTrackerServers() []TrackerServerPublic {
	return mtg.servers
}

type MockTrackerServerPublic struct {
	conn  ConnectionPublic
	index int
}

func (mts *MockTrackerServerPublic) GetConnection() ConnectionPublic {
	return mts.conn
}
func (mts *MockTrackerServerPublic) GetIndex() int {
	return mts.index
}

type MockConnectionPublic struct{}

func (mc *MockConnectionPublic) Close() error { return nil }

func TestConstructorsAndGetErrorCodePublic(t *testing.T) {
	g := &MockTrackerGroupPublic{}
	tc := NewTrackerClientPublic(g)
	assert.Equal(t, g, tc.tracker_group)
	tc2 := NewTrackerClientPublic(g)
	tc2.errno = 10
	assert.Equal(t, 10, tc2.GetErrorCode())
}

func TestGetTrackerServerPublic(t *testing.T) {
	server := &MockTrackerServerPublic{}
	group := &MockTrackerGroupPublic{server: server}
	tc := NewTrackerClientPublic(group)
	assert.Equal(t, server, tc.GetTrackerServer())
}

func TestGetConnectionFailoverPublic(t *testing.T) {
	server1 := &MockTrackerServerPublic{conn: nil, index: 0}
	server2 := &MockTrackerServerPublic{conn: &MockConnectionPublic{}, index: 1}
	group := &MockTrackerGroupPublic{servers: []TrackerServerPublic{server1, server2}}
	tc := NewTrackerClientPublic(group)
	assert.Equal(t, server2.conn, tc.GetConnection(nil))
}